registry/registry.go

```diff
@@ -205,51 +196,171 @@ func (r *dockerRegistry) doRequest(ctx context.Context, method, path string, hea
 		for k, v := range headers {
 			req.Header.Set(k, v)
 		}
-		for k, v := range authHeaders {
-			req.Header[k] = v
-		}
+		r.fillAuthCredentials(req)
 		resp, err = r.client.Do(req)
-		if err != nil {
-			if _, ok := err.(net.Error); ok {
-				continue
+		if _, ok := err.(net.Error); ok {
+			continue
+		}
+		if resp.StatusCode == http.StatusUnauthorized && resp.Header.Get("WWW-Authenticate") != "" && r.checkTokenIsValidForRenew() {
+			to := auth.TokenOptions{}
+			to.Username = r.authConfig.Username
+			to.Secret = r.authConfig.Password
+			challenges := auth.ParseAuthHeader(resp.Header)
+			for _, c := range challenges {
+				if c.Scheme == auth.BearerAuth {
+					to.Realm = c.Parameters["realm"]
+					to.Service = c.Parameters["service"]
+					to.Scopes = append(to.Scopes, c.Parameters["scope"])
+				}
+			}
+			to.Scopes = parseScopes(to.Scopes).normalize()
+			respAuth, tokenErr := auth.FetchToken(ctx, r.client, req.Header, to)
+			if tokenErr != nil {
+				return nil, tokenErr
+			}
+			if respAuth.IssuedAt.IsZero() {
+				respAuth.IssuedAt = time.Now()
+			}
+			if respAuth.ExpiresIn == 0 {
+				respAuth.ExpiresIn = defaultExpiration
 			}
+			if exp := respAuth.IssuedAt.Add(time.Duration(float64(respAuth.ExpiresIn)*0.9) * time.Second); time.Now().Before(exp) {
+				r.expires = exp
+			}
+			r.token = respAuth.Token
+			goto request
```

Comment:
@morpheu we dont have control to avoid loop of renew, we could implement a max retries